src/bootupd.rs

```diff
@@ -32,29 +23,27 @@ pub(crate) enum ClientRequest {
 }
 
 pub(crate) fn install(source_root: &str, dest_root: &str) -> Result<()> {
-    let statepath = Path::new(dest_root)
-        .join(STATEFILE_DIR)
-        .join(STATEFILE_NAME);
-    if statepath.exists() {
-        bail!("{:?} already exists, cannot re-install", statepath);
-    }
+    SavedState::ensure_not_present(dest_root)
+        .context("failed to install, invalid re-install attempted")?;
 
     let components = get_components();
     if components.is_empty() {
         println!("No components available for this platform.");
         return Ok(());
     }
-    let mut state = SavedState {
-        installed: Default::default(),
-        pending: Default::default(),
-    };
+
+    let mut state = SavedState::default();
     for component in components.values() {
         let meta = component.install(source_root, dest_root)?;
         state.installed.insert(component.name().into(), meta);
     }
 
-    let sysroot = openat::Dir::open(dest_root)?;
-    update_state(&sysroot, &state).context("Failed to update state")?;
+    let mut state_guard =
+        SavedState::acquire_write_lock(source_root).context("failed to acquire write lock")?;
```

Comment:
I'm not sure how much sense it makes though to grab any lock here.  The source root isn't being changed.

TBH I think the real bug here is that rpm-ostree is probably not mounting `/run` when it's running postprocessing scripts.

But for now...well, I guess if this works it doesn't hurt but I'd probably have said let's skip locking.